spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/ApiKeyValidityVerifier.java

```diff
@@ -0,0 +1,122 @@
+    private Timestamp adminsLocalTime = new Timestamp(0);
+
+    public boolean doesResultValidateApiKey(ResponseEntity<ApiKey> resultOfApiKeySearch, String apiKeyValue) {
+        if(apiKeyValue != null)
+            this.apiKeyValue = apiKeyValue;
+        else {
+            log.info("The API key value was null and could therefore not be checked for validity.");
+            return false;
+        }
+
+        if (resultOfApiKeySearch != null) {
+            return isApiKeyPresentAndValid(resultOfApiKeySearch);
+        } else {
+            log.info("The ResponseEntity is null. Therefore, the key is checked to be validated by " +
+                    "the local API key list.");
+            return isApiKeyValidInLocalApiKeyList();
+        }
+    }
+
+    private boolean isApiKeyPresentAndValid(ResponseEntity<ApiKey> resultOfApiKeySearch) {
+        ApiKey apiKey = resultOfApiKeySearch.getBody();
+        if (apiKey != null) {
+            log.debug("CoatRack admin found a valid key with the value: " + apiKeyValue);
+            return isApiKeyNeitherDeletedNorExpired(apiKey);
+        } else {
+            log.debug("API key value could not be found by CoatRack Admin: " + apiKeyValue);
+            return false;
+        }
+    }
+
+    private boolean isApiKeyNeitherDeletedNorExpired(ApiKey apiKey) {
+        boolean isExpired = apiKey.getValidUntil().getTime() < adminsLocalTime.getTime();
+        boolean isDeleted = apiKey.getDeletedWhen() != null;
+        logIfApiKeyIsExpiredOrDeleted(isExpired, isDeleted);
+        return !(isDeleted || isExpired);
+    }
+
+    private void logIfApiKeyIsExpiredOrDeleted(boolean isExpired, boolean isDeleted) {
+        String preamble = "Access to services denied. The api key with the value ";
+        String helpInstruction = "Please create a new one.";
+        if(isExpired)
+            log.info(preamble + apiKeyValue + " is expired. " + helpInstruction);
+        if(isDeleted)
+            log.info(preamble + apiKeyValue + " is deleted. " + helpInstruction);
+    }
+
+    private boolean isApiKeyValidInLocalApiKeyList() {
+        if (isApiKeyInLocalApiKeyList())
+            return wasLatestUpdateWithinTheLastHour();
+        else
+            return false;
+    }
+
+    private boolean isApiKeyInLocalApiKeyList() {
+        return apiKeyValueList.stream().anyMatch(x -> x.equals(apiKeyValue));
+    }
+
+    private boolean wasLatestUpdateWithinTheLastHour() {
+        return lastApiKeyValueListUpdate.getTime() + oneHourInMillis > System.currentTimeMillis();
+    }
+
+    public void setApiKeyList(List<String> apiKeyList){
+        this.apiKeyValueList = apiKeyList;
+    }
+
+    public void setLastApiKeyValueListUpdate(Timestamp lastApiKeyValueListUpdate) {
+        this.lastApiKeyValueListUpdate = lastApiKeyValueListUpdate;
+    }
+
+    public void setAdminsLocalTime(Timestamp adminsLocalTime) {
+        this.adminsLocalTime = adminsLocalTime;
```

Comment:
Is the admin's time at all relevant for the gateway? Gateway could be running in another timezone... Maybe it should only need to know the time itself last called admin?

Reply:
This adminsLocalTime feature is removed for now, but an issue will be created to discuss that matter.